### PR TITLE
Add a Kafka 0.10+ compatible collector and attempt to integrate into zipkin-server

### LINKED ISSUE #1405: Update Kafka (to 0.10)
Last year, we attempted to update Kafka to 0.9, but had to revert it due to the ordering problem implied.

https://github.com/openzipkin/zipkin/pull/904

Unfortunately no one responded to the email asking about kafka update plans: https://groups.google.com/forum/#!topic/zipkin-user/rQpecbvc6Y0

We shouldn't be pinned to 0.8 forever, and we are now getting requests for 0.10 support. I'm wondering how we should address Kafka updates, particularly as the server includes an all-jar. Unless packaging is different, it might imply weird tricks to be able support multiple versions in the same binary, or a separate build.

Can anyone share their current status wrt kafka and/or have ideas on how to support multiple versions simply?

cc @prat0318 @dgarson @SimenB @sveisvei @eirslett @kristofa @NegatioN @shakuzen

### PR DESCRIPTION
**Pull request by dgrabows**
These changes add a new collector module supporting Kafka 0.10+ (zipkin-collector-kafka10) and the associated auto-configure module (zipkin-autoconfigure-collector-kafka10). They also make the auto-configure module a dependency of zipkin-server, with the collector activated by setting the `KAFKA_BOOSTRAP_SERVERS` environment variable. See https://github.com/dgrabows/zipkin/tree/kafka10-collector/zipkin-server#kafka-010-collector for more details on the supported configuration.

Further changes are needed to how the new collector is integrated with zipkin-server. As currently implemented, the 0.10.2.0 version of the org.apache.kafka:kafka-clients dependency takes precendence over the 0.8.x.x one when zipkin-server is built. This will prevent the existing 0.8 Kafka collector from working. From https://github.com/openzipkin/zipkin/issues/1405#issuecomment-300023053:

> wrt how we address integration, I think the best way is to act like sqs,
where we package a module jar (and this includes a zipkin-server-kafka file
like
https://github.com/openzipkin/zipkin-aws/blob/master/autoconfigure/collector-sqs/src/main/resources/zipkin-server-sqs.yml
)
>
> The pom would include a module file as well, like this, except different
excludes:
https://github.com/openzipkin/zipkin-aws/blob/master/autoconfigure/collector-sqs/pom.xml
>
> Running will be a little messier than this because of classpath
https://github.com/openzipkin/zipkin-aws/tree/master/autoconfigure/collector-sqs#running
>
> For example, we'll probably have to customize the docker image based on
presence of the KAFKA_BOOTSTRAP_SERVERS variable, and move old jars out of
the way.. we may end up lucky enough to not need to do that. Basically,
once there's a module dir the next step is seeing how lucky we are.. in
best case classpath order works and we don't need to hack the all jar to
take out the kafka08 stuff.
>
> In the future, we could make a thin launcher
https://github.com/dsyer/spring-boot-thin-launcher as the default image

Outstanding tasks I'm aware of:
- [x] Add unit test to https://github.com/dgrabows/zipkin/blob/kafka10-collector/zipkin-collector/kafka10/src/test/java/zipkin/collector/kafka10/KafkaCollectorTest.java#L204 for collector running multiple threads.
- [x] Review/update closing of Kafka consumers when consumer thread is interrupted. See https://github.com/dgrabows/zipkin/blob/kafka10-collector/zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaConsumerProcessor.java#L83
- [x] Add zipkin-server-kafka.yml configuration file to https://github.com/dgrabows/zipkin/tree/kafka10-collector/zipkin-autoconfigure/collector-kafka10/src/main/resources
- [x] Update https://github.com/dgrabows/zipkin/blob/kafka10-collector/zipkin-autoconfigure/collector-kafka10/pom.xml to produce a module jar
- [x] Remove zipkin-autoconfigure-collector-kafka10 dependency from https://github.com/dgrabows/zipkin/blob/kafka10-collector/zipkin-server/pom.xml
- [x] Move usage documentation from https://github.com/dgrabows/zipkin/blob/kafka10-collector/zipkin-server/README.md#kafka-010-collector to new README.md in https://github.com/dgrabows/zipkin/tree/kafka10-collector/zipkin-autoconfigure/collector-kafka10 and update to reflect approach of loading the kafka10 collector from an external module jar
- [x] Uncomment https://github.com/dgrabows/zipkin/blob/kafka10-collector/zipkin-autoconfigure/collector-kafka10/src/test/java/zipkin/collector/kafka10/ZipkinKafkaCollectorAutoConfigurationTest.java#L88 and get test passing
- [x] Clean up logging APIs, implementations, and configuration in https://github.com/dgrabows/zipkin/tree/kafka10-collector/zipkin-collector/kafka10

There are probably also changes to https://github.com/openzipkin/docker-zipkin that are not yet clear to me, although that's probably a separate pull request given that it's a different repository.

Attempts to resolve #1405.

### CODE REVIEW

**Comment by dgrabows** [comment]
The zipkin-autoconfigure/collector-kafka10 module now builds a module jar, e.g. `zipkin-autoconfigure-collector-kafka10-1.24.1-SNAPSHOT-module.jar`.

Next step is getting the collector to run in zipkin-server again as an external module, based off of the SQS collector example and instructions on https://github.com/openzipkin/zipkin/issues/1219#issuecomment-274451253.

**Comment by codefromthecrypt** [comment]
Thanks for getting this started!

Here are some notes as most looks picture perfect!

* toss the extra logback config in test sources. Add a test dep of a slf4j
to log4j2 bridge if somehow missing
* if slf4j is the implied logging api of kafka, make a comment in the pom
of why we are binding to it. Otherwise switch to whatever kafka10 uses and
similarly make a comment
* you have a commented out test. You should be able to test this by
sneakily using package access to the kafkaConsumers field. Try it!
* move/rework the kafka properties description from zipkin-server to the
autoconfig README. you can look at zipkin-sparkstreaming or zipkin-aws for
exampkes.. mainly we should link in nice bold in the zipkin-server readme
to the right place. (This is different than othe4s but a good time to break
from sprawling)

All above are small things.. good work!

**Comment by dgrabows** [comment]
I made some changes to clean up logging for the unit tests, but I'm not sure if I went in the direction you intended. 

I took your comments to mean there is a preference for binding the logging to a log4j implementation, although that seems to differ from the preference in elsewhere in the project. Modules that reference a logging implementation all seem to depend on logback and https://github.com/openzipkin/zipkin/blob/master/pom.xml defines a logback.version. I can't find any references to log4j other than some dependency exclusions and the unnecessary log4j.properties in test/resources for zipkin-collector-kafka, which I carried over when I created the kafka10 collector.

org.apache.kafka:kafka-clients, at least for 0.10+, implement logging against slf4j-api.

I opted to delete the log4j.properties and do some cleanup on the logback.xml. I can easily go the other direction and swap logback for slf4j, but this seemed more consistent with other modules. Just let me know.

**Comment by codefromthecrypt** [comment]
>
> I made some changes to clean up logging for the unit tests, but I'm not
> sure if I went in the direction you intended.
>
this is fine. I forgot we hadn't switched tests in the main repo to log4j2,
yet. my mistake!

**Comment by dgrabows** [comment]
The kafka10 collector works with zipkin-server as an external module jar. There was no classpath manipulation required. I pretty much just worked once I figured out how to get all the right stuff into `loader.path`. See https://github.com/dgrabows/zipkin/tree/kafka10-collector/zipkin-autoconfigure/collector-kafka10 for instructions on how to run it.

**Comment by codefromthecrypt** [comment]
great. do you want to try and get conditional loading working in the docker
image? There might be a chicken/egg problem as the docker image pulls
release images..

**Comment by dgrabows** [comment]
I spent some time working to update the docker image today. Most of that was spent getting familiar with docker, docker-compose, etc., as I've not done much more than copy/paste execution of containers previously. I'm hoping tomorrow I can get something hacked together and into a PR in docker-zipkin so I can get some early feedback.

**Comment by StephenWithPH** [comment]
@dgrabows let me know if I can do anything to help with the Docker work.

**Comment by dgrabows** [comment]
It took a lot longer than I was hoping, but I've got something working with the docker images.

See https://github.com/dgrabows/docker-zipkin/tree/kafka10-collector and the comments on the commit https://github.com/dgrabows/docker-zipkin/commit/02470624be97b2b5108b91c2b138be96c1a7d160.

I don't expect any of that to be in a final state, but it's a (more or less) working starting point for discussion on how this should work, what to do about upgrading the kafka container image to 0.10, and how to deal with differences between (at least on Macs) Docker Toolbox and Docker for Mac.

I'll try to write up a cohert PR tonight for those changes in docker-zipkin to allow for more detailed feedback.

@StephenWithPH Any feedback on the direction those docker changes are going would be greatly appreciated.

**Comment by StephenWithPH** [comment]
@dgrabows looking forward to your PR in docker-zipkin. In the interest of keeping the discussion contained, I'll add comments there. I think everything is very close. Thank you for taking this on!

**Comment by dgrabows** [comment]
Pull request for the initial attempt a docker changes is now created.

https://github.com/openzipkin/docker-zipkin/pull/142

**Comment by codefromthecrypt** [comment]
Thanks. How do you feel about this PR btw. Anything left?

**Comment by dgrabows** [comment]
That last commit address my final outstanding todo. Assuming the tests pass (they did for me locally) and there are no concerns from anyone else, I think this is good to go.

File: zipkin-autoconfigure/collector-kafka10/README.md
Comment (codefromthecrypt) [inline zipkin-autoconfigure/collector-kafka10/README.md]: is JRE8 a Kafka 0.10 constraint? If not, I might drop it back to JRE 7 as some use the components directly, and we usually don't use a JRE level higher than the underlying driver.

File: zipkin-autoconfigure/collector-kafka10/README.md
Comment (codefromthecrypt) [inline zipkin-autoconfigure/collector-kafka10/README.md]: remember some use this code w/o autoconfigure (as a library). Notes about autoconfigure and modules etc ideally go into the auto-configure module README, since it is produced there. You can look at this for an example:

https://github.com/openzipkin/zipkin-aws/tree/master/collector-sqs
https://github.com/openzipkin/zipkin-aws/tree/master/autoconfigure/collector-sqs

File: zipkin-autoconfigure/collector-kafka10/pom.xml
Comment (codefromthecrypt) [inline zipkin-autoconfigure/collector-kafka10/pom.xml]: note: rebase and bump versions as we are now at 1.25.1-SNAPSHOT

File: zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/KafkaBootstrapServersSetCondition.java
Comment (codefromthecrypt) [inline zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/KafkaBootstrapServersSetCondition.java]: I've started inlining little classes like this as static inner classes on the autoconfigure type. It keeps the header-weight down, and they are pretty insignificant.

Also, for single-property conditions, it is unambiguous what failed, so you can use the slightly less complex normal condition as `@Conditional(KafkaBootstrapServersSet.class)`:

ex.
```java
  static final class HttpLoggingSet implements Condition {
    @Override public boolean matches(ConditionContext condition, AnnotatedTypeMetadata ignored) {
      return !isEmpty(condition.getEnvironment()
          .getProperty("zipkin.storage.elasticsearch.http-logging"));
    }
  }
```

File: zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorAutoConfiguration.java
Comment (codefromthecrypt) [inline zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorAutoConfiguration.java:40]: ex this is a tiny file, you can add condition type here (noting you'll need to qualify it )

File: zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java
Comment (codefromthecrypt) [inline zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java]: Latest fashion is to provide one-line docs here `/** password used for basic auth. Needed when Shield or X-Pack security is enabled */` and leave all values at null (imply change streams to Integer).

The one-lines shouldn't include javadoc tags like `@see` as this is looked at by spring boot utilities that add a tool tip.

File: zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java
Comment (codefromthecrypt) [inline zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java:74]: this doesn't need to be public as nothing else uses it.

File: zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java
Comment (codefromthecrypt) [inline zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java]: here, you can use this form to avoid accidentally overriding defaults (which avoids coordinating defaults in two places)

```
    KafkaCollector.Builder result = KafkaCollector.builder();
    if (topic != null) result.topic(topic);
    if (groupId != null) result.groupId(groupId);
    if (bootstrapServers != null) result.bootstrapServers(bootstrapServers);
   ...
```

File: zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java
Comment (codefromthecrypt) [inline zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java:54]: note: you'll want to coerce empty to null as sometimes people can't set null with properties. Usually I add a static method for this called emptyToNull

File: zipkin-autoconfigure/collector-kafka10/src/test/java/zipkin/collector/kafka10/ZipkinKafkaCollectorAutoConfigurationTest.java
Comment (codefromthecrypt) [inline zipkin-autoconfigure/collector-kafka10/src/test/java/zipkin/collector/kafka10/ZipkinKafkaCollectorAutoConfigurationTest.java:34]: here's a neat way to test properties set properly. you'll still need a test to ensure the object only loads when its signaling property is set https://github.com/openzipkin/zipkin-sparkstreaming/blob/master/autoconfigure/stream-kafka/src/test/java/zipkin/sparkstreaming/autoconfigure/stream/kafka/ZipkinKafkaStreamFactoryPropertiesTest.java

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollector.java
Comment (codefromthecrypt) [inline zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollector.java:117]: while not required, you could consider using autovalue to make your builder https://github.com/openzipkin/zipkin-sparkstreaming/blob/master/stream/kafka/src/main/java/zipkin/sparkstreaming/stream/kafka/KafkaStreamFactory.java

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollector.java
Comment (codefromthecrypt) [inline zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollector.java]: I think we usually wait 1 second by default now.. though it is arbitrary anyway

File: zipkin-collector/kafka10/src/test/java/zipkin/collector/kafka10/KafkaCollectorTest.java
Comment (codefromthecrypt) [inline zipkin-collector/kafka10/src/test/java/zipkin/collector/kafka10/KafkaCollectorTest.java]: maybe pull this into a method named warmUp with this as its javadoc?

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java
Comment (codefromthecrypt) [inline zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java]: nit: probably looks nicer to just accept the Builder and take values from it (using package protected)

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java
Comment (codefromthecrypt) [inline zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java]: doesn't need to be public. you can use the package protected field for testing. that lets you move this javadoc to the field itself.

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java
Comment (codefromthecrypt) [inline zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java]: hmm ok :)

File: zipkin-server/README.md
Comment (codefromthecrypt) [inline zipkin-server/README.md]: move this to a note under "This collector remains a Kafka 0.8.x consumer, while Zipkin systems update to 0.9+."? I think more people will notice it this way.

**Review by codefromthecrypt — APPROVED** [review]
I made mostly style comments, in efforts to make this collector have all the latest practice. That way, the next to copy/paste will have a better starting point.

The code itself is great. Really appreciate the tests. Please coordinate with @fedj and he can merge this when ready. Then, this will be published into the jfrog snapshots repository. Once that's in place, you can adjust your pending docker PR to download from there, which makes things more realistic. Once that's all together, coordinate w/ @fedj for a release of zipkin. After the zipkin release, you can adjust the docker PR to work against the release repository. Once that's in, it can be merged and released (or the version can be re-released). Please let me know any questions on this

File: zipkin-autoconfigure/collector-kafka10/README.md
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/README.md]: This is the autoconfigure readme. :)

I can change the title of the page to make that clearer.

File: zipkin-autoconfigure/collector-kafka10/README.md
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/README.md]: JRE7 is the minimum requirement for using kafka-clients, including version 0.10. I copied the JRE8 requirement from the zipkin-server readme.

There's no reason collector-kafka10 can't be JRE 7 compatible. I'll check the code to see if there's anything Java 8 specific (other than labmdas, which I'm assuming retrolambda will take care of).

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java
Comment (dgrabows) [inline zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java]: I think I'll need to replace the forEach() call here to maintain Java 7 compatibility.

File: zipkin-autoconfigure/collector-kafka10/README.md
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/README.md]: There is one use of forEach() in KafkaCollectorWorker that needs to be replaced (see comment further down).

**Comment by codefromthecrypt** [comment]
>
> I think I'll need to replace the forEach() call here to maintain Java 7
> compatibility.
>
cool. I think one for loop is reasonable. If it had higher impact, I'd
recommend retrolambda

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java
Comment (dgrabows) [reply zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java]: I went back and forth on this a few times. It seems awkard. But, without it, the Kafka consumers will rarely get closed gracefully, because a thread interrupt is the primary reason the polling loop will be terminated and the kafka consumer will throw an exception if the thread has the interrupted status set.

Closing the consumer isn't strictly necessary, although it can reduce the delay before the consumer group rebalances to account for the consumer being gone.

Your reaction is enough to tip me towards taking it out, since it was a 50/50 decision to begin with. I will do that.

File: zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorAutoConfiguration.java
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorAutoConfiguration.java:40]: Done.

File: zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java:54]: Done.

File: zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java:74]: I ended up leaving this public, because ZipkinKafkaCollectorPropertiesOverrideTest needed access to both this method and the package-private fields on KafkaCollector.Builder. This seemed better than adding public getters for each of those fields.

File: zipkin-autoconfigure/collector-kafka10/src/test/java/zipkin/collector/kafka10/ZipkinKafkaCollectorAutoConfigurationTest.java
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/src/test/java/zipkin/collector/kafka10/ZipkinKafkaCollectorAutoConfigurationTest.java:34]: Done in newly added ZipkinKafkaCollectorPropertiesOverrideTest.

File: zipkin-autoconfigure/collector-kafka10/README.md
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/README.md]: I updated this to clarify that zipkin-server and the kafka10 collector have difference java/JRE requirements.

File: zipkin-autoconfigure/collector-kafka10/README.md
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/README.md]: I did change the page title and make some other edits to this readme and the readme for the collector module to try and clarify this.

File: zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/KafkaBootstrapServersSetCondition.java
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/KafkaBootstrapServersSetCondition.java]: Done.

File: zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java]: Done.

File: zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/src/main/java/zipkin/autoconfigure/collector/kafka10/ZipkinKafkaCollectorProperties.java]: Done.

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollector.java
Comment (dgrabows) [reply zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollector.java]: I switched this to wait 1 second.

File: zipkin-collector/kafka10/src/test/java/zipkin/collector/kafka10/KafkaCollectorTest.java
Comment (dgrabows) [reply zipkin-collector/kafka10/src/test/java/zipkin/collector/kafka10/KafkaCollectorTest.java]: Done.

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java
Comment (dgrabows) [reply zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java]: Done.

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java
Comment (dgrabows) [reply zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java]: Moved the comment and removed this method.

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java
Comment (dgrabows) [reply zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java]: The call to Thread.interrupted() is gone now.

File: zipkin-server/README.md
Comment (dgrabows) [reply zipkin-server/README.md]: Done.

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java
Comment (dgrabows) [reply zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollectorWorker.java]: Done.

File: zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollector.java
Comment (dgrabows) [reply zipkin-collector/kafka10/src/main/java/zipkin/collector/kafka10/KafkaCollector.java:117]: I ran out of steam before getting to this. If anybody feels strongly about using autovalue here, let me know and I will circle back around to it.

**Comment by codefromthecrypt** [comment]
>
> I ran out of steam before getting to this. If anybody feels strongly about
> using autovalue here, let me know and I will circle back around to it.
>
not important!

File: zipkin-autoconfigure/collector-kafka10/pom.xml
Comment (dgrabows) [reply zipkin-autoconfigure/collector-kafka10/pom.xml]: I've rebased the kafka10-collector branch in my local repo and committed the version updates to these two pom files. It won't push cleanly, because my local branch and the one on github have diverged. I suspect I can force the push to resolve this (`git push -f origin +kafka10-collector'), but I'm wary of doing something that rewrites history and screws up the PR.

Am I heading in the right direction, or did I not execute the rebase properly?

Current state of affairs on my local clone of the repo:
```dgrabowski:zipkin$ git status
On branch kafka10-collector
Your branch and 'origin/kafka10-collector' have diverged,
and have 37 and 25 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
nothing to commit, working tree clean
dgrabowski:zipkin$ git push
To github.com:dgrabows/zipkin.git
 ! [rejected]        kafka10-collector -> kafka10-collector (non-fast-forward)
error: failed to push some refs to 'git@github.com:dgrabows/zipkin.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
dgrabowski:zipkin$```

**Comment by codefromthecrypt** [comment]
I don't think anyone else is using your branch, so you can force push it.
The commits will be squashed on merge anyway.

**Comment by dgrabows** [comment]
Alright. Rebase and version catch-up are complete. Assuming the tests pass like they did locally, I think that's all the outstanding items.

**Comment by codefromthecrypt** [comment]
ok will merge on green, which would then put it into the snapshot repo

**Comment by codefromthecrypt** [comment]
Grand work, btw 👍

**Comment by dgrabows** [comment]
Thank you.

**Pull request merged**
